packages/application-shell/src/components/navbar/use-navbar-state-manager.ts

```diff
@@ -0,0 +1,197 @@
+import React from 'react';
+import isNil from 'lodash/isNil';
+import { useQuery } from 'react-apollo';
+import { useFeatureToggle } from '@flopflip/react-broadcast';
+import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
+import { TApplicationContext } from '@commercetools-frontend/application-shell-connectors';
+import { reportErrorToSentry } from '@commercetools-frontend/sentry';
+import { TApplicationsMenu, TNavbarMenu } from '../../types/generated/proxy';
+import {
+  TFetchProjectExtensionsNavbarQuery,
+  TFetchProjectExtensionsNavbarQueryVariables,
+} from '../../types/generated/settings';
+import { PROJECT_EXTENSIONS } from '../../feature-toggles';
+import { STORAGE_KEYS } from '../../constants';
+import useApplicationsMenu from '../../hooks/use-applications-menu';
+import FetchProjectExtensionsNavbar from './fetch-project-extensions-navbar.settings.graphql';
+import nonNullable from './non-nullable';
+
+type HookProps = {
+  environment: TApplicationContext<{}>['environment'];
+  DEV_ONLY__loadNavbarMenuConfig?: () => Promise<TApplicationsMenu['navBar']>;
+};
+
+const useNavbarStateManager = (props: HookProps) => {
+  const navBarNode = React.useRef<HTMLElement>(null);
+  const areProjectExtensionsEnabled = useFeatureToggle(PROJECT_EXTENSIONS);
+  const applicationsNavBarMenu = useApplicationsMenu<'navBar'>('navBar', {
+    queryOptions: {
+      onError: reportErrorToSentry,
+    },
+    skipRemoteQuery: !props.environment.servedByProxy,
+    loadMenuConfig: props.DEV_ONLY__loadNavbarMenuConfig,
+  });
+  const { data: projectExtensionsQuery } = useQuery<
+    TFetchProjectExtensionsNavbarQuery,
+    TFetchProjectExtensionsNavbarQueryVariables
+  >(FetchProjectExtensionsNavbar, {
+    skip: !props.environment.servedByProxy || !areProjectExtensionsEnabled,
+    variables: {
+      target: GRAPHQL_TARGETS.SETTINGS_SERVICE,
+    },
+    fetchPolicy: 'cache-and-network',
+    onError: reportErrorToSentry,
+  });
+  const customAppsMenu =
+    projectExtensionsQuery &&
+    projectExtensionsQuery.projectExtension &&
+    projectExtensionsQuery.projectExtension.applications
```

Comment:
I know it's an obvious question but can't TypeScript recently optionally chain into it?

Reply:
True, will change thanks :)

Reply:
https://github.com/commercetools/merchant-center-application-kit/pull/1273/commits/ccf8cc5bea3ba9edff296057a6e49d7fe19e0be5

Reply:
Mäh, we don't use the babel plugin yet.

<img width="1177" alt="image" src="https://user-images.githubusercontent.com/1110551/73127893-917ab000-3fc7-11ea-9649-f35f14152e18.png">

Will keep it like this for now and look into enabling optional chaining in another PR.